consensus/integration_test.go

```diff
@@ -0,0 +1,247 @@
+	genesisBlock := createDummyBlockWithVoteQC(createDummyBlockHash(), 2, 2, allNodes)
+
+	// Set the crank and timeout durations
+	crankTimer := time.Millisecond * 500    // Produce a block every 500ms
+	timeoutTimer := time.Millisecond * 1000 // Time out if there is no block broadcast within 1000ms
+
+	// Initialize all nodes and and connect them to each other
+	for _, validator := range allNodes {
+		require.NoError(t, validator.Init(crankTimer, timeoutTimer, genesisBlock, allNodes))
+	}
+
+	// Start all nodes' consensus event loops.
+	for _, node := range allNodes {
+		node.Start()
+	}
+
+	// Broadcast a block with a valid QC to all nodes. This kicks off the steady state flow
+	// of the Fast-HotStuff consensus. All nodes will begin voting starting from this block.
+	initialBlockToBroadcast := createDummyBlockWithVoteQC(genesisBlock.GetBlockHash(), 3, 3, allNodes)
+	for _, node := range allNodes {
+		node.ProcessBlock(initialBlockToBroadcast)
+	}
+
+	// Let all nodes run for 1 second so that the network produces at least one block.
+	time.Sleep(time.Millisecond * 1000)
+
+	// Stop node 3 to simulate the network going down catastrophically. All nodes will begin to
+	// time out. No blocks will be proposed during this time.
+	node3.Stop()
+
+	// Let all online nodes run for 2 seconds so they time out at least once. The block proposer
+	// may proposer at most one block during this time if it had enough votes stored for its current
+	// tip to build a QC. The QC will not reach node 3.
+	//
+	// The network has halted during this period.
+	time.Sleep(time.Millisecond * 2000)
+
+	// Cache node 3's tip during the network halt. Nodes internal clocks and view will during this
+	// period.
+	node3TipDuringNetworkHalt := node3.eventLoop.tip.block.(*block)
+
+	// After a catastrophic network failure, nodes with a super-majority of stake need to somehow
+	// agree on a starting state of the chain. As long as they are able to sync from or to any peer
+	// that eventually becomes a block proposer, the network will recover.
+	node3.Resync(node1.eventLoop.tip.block.(*block), collections.MapValues(node1.safeBlocks))
+	node3.Start()
+
+	// Let all nodes run for 10 seconds. Eventually all nodes will converge on a single view and
+	// the network will start producing blocks again.
+	time.Sleep(time.Millisecond * 10000)
+
+	// Stop remaining nodes
+	for _, node := range allNodes {
+		node.Stop()
+	}
+
+	// Validate and print resulting chain of blocks stored by all nodes
+	validateAndPrintBlockChain(t, node1, node1.safeBlocks)
+	validateAndPrintBlockChain(t, node2, node2.safeBlocks)
+	validateAndPrintBlockChain(t, node3, node3.safeBlocks)
+
+	// Verify that the network has produced at least one block since it recovered
+	require.Greater(t, node3.eventLoop.tip.block.GetView(), node3TipDuringNetworkHalt.GetView())
+	require.Greater(t, node3.eventLoop.tip.block.GetHeight(), node3TipDuringNetworkHalt.GetHeight())
+}
```

Comment:
This whole framework seems very solid for integration testing. When you combine it with the unit testing you're doing in event_loop_test.go, it feels like we have everything we need to exercise and bullet-proof everything here.

Is there anything you want to add in terms of testing infrastructure beyond this in the short-term?

One thing I thought of is maybe you'd want to be able to "fast-forward" time so that you don't have to wait for the wall clock to hit. But this comes with its own challenges, and it's not obviously worth it because it isn't as "real" as what you're doing here. Still maybe worth looking into whether this is a "thing" that people do in Go integration tests ever.

Reply:
I want to test different stake distributions across nodes and downtime durations for nodes. The current framework basically allows us to test any permutation of nodes, so I want to stress test as much as possible before we make it out to devnet. It would be somewhere on the order of 3-5 more integration tests similar to what already exists.

The "fast-forward" time comment is interesting. It would allow us to test scenarios with extremely consecutive timeouts in the network. I'll look into it.

Reply:
Just an update here that we have near complete coverage of unit test cases, and a robust sequence of integration tests for the event loop here. At this point, we're best off spending testing effort on the production Server.go.